### ISSUE TITLE
김민지

### ISSUE THREAD

**Issue by minji-kim5** [seongjun25/project_diecasting#5]
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

train=pd.read_csv('./train.csv')
train.head()
train.info()

### [upper_mold_temp1,upper_mold_temp2,upper_mold_temp3,lower_mold_temp1,lower_mold_temp2,lower_mold_temp3] 6개 열 분석

df=train.iloc[:,17:23]
df.info()
df.isnull().sum()
df.describe()

num_cols=df.select_dtypes(include='number').columns.to_list()
cat_cols=df.select_dtypes(include='object').columns.to_list()

import matplotlib.pyplot as plt
import seaborn as sns

# 불량과 정상 제품 그룹으로 나누기
grouped = train.groupby('passorfail')

# 불량 그룹 (passorfail == 1)과 정상 그룹 (passorfail == 0)로 나누기
df_defective = df[train['passorfail'] == 1]
df_normal = df[train['passorfail'] == 0]

# 비교할 변수들
temp_cols = ['upper_mold_temp1', 'upper_mold_temp2', 'upper_mold_temp3',
             'lower_mold_temp1', 'lower_mold_temp2', 'lower_mold_temp3']

# -----------------------------------------------------

### 상관관계 분석 -> 불량에 영향 주는 파라미터 추출
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from scipy.stats import pointbiserialr

# 불량 여부 컬럼명
features = train.drop(columns='passorfail').select_dtypes(include='number').columns

# 1. Point-Biserial Correlation (이진라벨과 연속형 변수 상관)
train_clean = train[features].copy()

train_clean = train_clean.dropna()
train_clean['passorfail'] = train['passorfail']
correlations = {}
for col in features:
    corr, _ = pointbiserialr(train_clean['passorfail'], train_clean[col])
    correlations[col] = corr

# 상관계수 시각화
corr_series = pd.Series(correlations).sort_values(key=abs, ascending=False)
print("불량과의 상관관계 (절대값 기준 정렬):")
print(corr_series)

sns.barplot(x=corr_series.values, y=corr_series.index)
plt.title('Point-Biserial Correlation with Defect')
plt.xlabel('Correlation')
plt.ylabel('Feature')
plt.show()

## 2. 랜덤포레스트로 피처 중요도
X = train[features]
y = train['passorfail']
X = X.fillna(X.mean())
# 스케일링 (선택사항)
X_scaled = StandardScaler().fit_transform(X)

model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_scaled, y)

importances = pd.Series(model.feature_importances_, index=features).sort_values(ascending=False)

print("\n랜덤포레스트 기반 중요도:")
print(importances)

sns.barplot(x=importances.values, y=importances.index)
plt.title('Feature Importance (Random Forest)')
plt.xlabel('Importance')
plt.ylabel('Feature')
plt.show()

# 랜덤포레스트 기반 중요도: 0.05 이상이면 관련성 높음!!

# 박스 플롯으로 각 변수의 분포 비교
import seaborn as sns
import matplotlib.pyplot as plt
plt.figure(figsize=(15, 10))

for i, col in enumerate(temp_cols, 1):
    plt.subplot(2, 3, i)
    sns.boxplot(x='passorfail', y=col, data=train)
    plt.title(f'{col} Distribution by Pass/Fail')
    plt.xlabel('Pass/Fail (0: Normal, 1: Defective)')
    plt.ylabel(f'{col} Temperature')

plt.tight_layout()
plt.suptitle('Temperature Distribution by Pass/Fail', fontsize=16, y=1.03)
plt.show()

# 이상치 -> mold_code와 관련 있는지 확인
train.loc[train['upper_mold_temp3']<=300,'mold_code'].value_counts()
train.loc[train['lower_mold_temp1']>=300,'mold_code'].value_counts()
train.loc[train['lower_mold_temp2']>=290,'mold_code'].value_counts()
train.loc[train['lower_mold_temp2']<=100,'mold_code'].value_counts()

# 극단치 제거
train[train['upper_mold_temp1']>=1400]   #1개/불량
train[train['upper_mold_temp2']>=400]    #1개/불량
train[train['lower_mold_temp3']>=60000]  #1개/양품

### 히스토그램으로 각 변수의 분포 비교
plt.figure(figsize=(15, 10))

for i, col in enumerate(temp_cols, 1):
    plt.subplot(2, 3, i)
    sns.histplot(df[train['passorfail'] == 0][col], bins=30, color='blue', label='Normal', kde=True)
    sns.histplot(df[train['passorfail'] == 1][col], bins=30, color='red', label='Defective', kde=True)
    plt.legend()
    plt.title(f'{col} Histogram by Pass/Fail')
    plt.xlabel(f'{col} Temperature')
    plt.ylabel('Frequency')

plt.tight_layout()
plt.suptitle('Temperature Histogram by Pass/Fail', fontsize=16, y=1.03)
plt.show()

### 막대 히스토그램
fig, axes = plt.subplots(nrows=len(num_cols), figsize=(10, 6 * len(num_cols)))

for i, col in enumerate(num_cols):
    sns.histplot(data=train, x=col, hue="passorfail", multiple="stack", kde=False, palette={0: 'skyblue', 1: 'sandybrown'}, ax=axes[i])
    axes[i].set_title(f'{col}')
# 그래프 레이아웃 조정
plt.tight_layout()
# 그래프 출력
plt.show()

# ------------------------------------------------------
###  [mold code별 box plot]

import seaborn as sns
import matplotlib.pyplot as plt

# mold_code별로 반복
unique_molds = train['mold_code'].unique()

# temp_cols는 기존 코드에서 정의한 것 사용
temp_cols = ['upper_mold_temp1', 'upper_mold_temp2', 'upper_mold_temp3',
             'lower_mold_temp1', 'lower_mold_temp2', 'lower_mold_temp3']

# mold_code별 boxplot 시각화
for mold in unique_molds:
    mold_df = train[train['mold_code'] == mold]
    
    plt.figure(figsize=(15, 10))
    
    for i, col in enumerate(temp_cols, 1):
        plt.subplot(2, 3, i)
        sns.boxplot(x='passorfail', y=col, data=mold_df)
        plt.title(f'{col} by Pass/Fail (Mold Code: {mold})')
        plt.xlabel('Pass/Fail (0: Normal, 1: Defective)')
        plt.ylabel('Temperature')

    plt.suptitle(f'Temperature Distribution by Pass/Fail - Mold Code: {mold}', fontsize=16, y=1.03)
    plt.tight_layout()
    plt.show()

# [mold code 별 히스토그램]
import seaborn as sns
import matplotlib.pyplot as plt

train.columns = train.columns.str.strip()
# mold_code를 문자열로 변환
train['mold_code'] = train['mold_code'].astype(str)

# 시각화 대상 온도 변수
temp_cols = ['upper_mold_temp1', 'upper_mold_temp2', 'upper_mold_temp3',
             'lower_mold_temp1', 'lower_mold_temp2', 'lower_mold_temp3']

## 히스토그램 그리기
for temp_col in temp_cols:
    plt.figure(figsize=(10, 6))
    sns.histplot(
        data=train,
        x=temp_col,
        hue='mold_code',
        element='step',
        stat='density',
        common_norm=False,
        palette='Set1',
        linewidth=1.5
    )
    plt.title(f'Distribution of {temp_col} by Mold Code')
    plt.xlabel('Temperature')
    plt.ylabel('Density')
    plt.legend(title='Mold Code')
    plt.tight_layout()
    plt.show()
